v2/common/src/main/java/com/google/cloud/teleport/v2/cdc/dlq/FileBasedDeadLetterQueueReconsumer.java

```diff
@@ -92,28 +93,39 @@ public PCollection<String> expand(PBegin in) {
         .apply("TriggerConsumeDLQ", GenerateSequence.from(0).withRate(1, this.recheckPeriod))
         .apply("AsFilePattern", MapElements.into(TypeDescriptors.strings()).via(seq -> filePattern))
         .apply("MatchFiles", FileIO.matchAll())
-        .apply("ConsumeMatches", moveAndConsumeMatches());
+        .apply("MoveAndConsumeMatches", moveAndConsumeMatches());
   }
 
-  /** Build a {@link PTransform} that consumes matched DLQ files. */
+  /**
+   * Build a {@link PTransform} that consumes matched DLQ files.
+   *
+   * <p>This transform is not idempotent. It should only be used when the source is guaranteed to
+   * provide each file only once.
+   */
   static PTransform<PCollection<Metadata>, PCollection<String>> moveAndConsumeMatches() {
     return new PTransform<PCollection<Metadata>, PCollection<String>>() {
       @Override
       public PCollection<String> expand(PCollection<Metadata> input) {
-        // TODO(pabloem, dhercher): Use a Beam standard transform once possible
-        // TODO(pabloem, dhercher): Add a _metadata attribute to track whether a row comes from DLQ.
         TupleTag<String> fileContents = new TupleTag<String>();
-        TupleTag<Metadata> fileMetadata = new TupleTag<Metadata>();
+        TupleTag<ResourceId> fileMetadata = new TupleTag<ResourceId>();
+
+        PCollection<ResourceId> movedFiles =
+            input
+                .apply("MoveFiles", ParDo.of(new MoveFiles()))
+                .setCoder(ResourceIdCoder.of())
+                .apply("ReshuffleMovedFiles", Reshuffle.viaRandomKey());
+
         PCollectionTuple results =
-            input.apply(
+            movedFiles.apply(
+                "ConsumeFiles",
                 ParDo.of(new MoveAndConsumeFn(fileContents, fileMetadata))
                     .withOutputTags(fileContents, TupleTagList.of(fileMetadata)));
 
         results
             .get(fileMetadata)
-            .setCoder(MetadataCoder.of())
+            .setCoder(ResourceIdCoder.of())
```

Comment:
I think this is inevitable, but we are breaking update compatibility. Maybe worth calling out in the PR title so that it gets into the resource notes.